### PR TITLE
Replace closures_captures and upvar_capture with closure_min_captures

### LINKED ISSUE #18: Use min capture results everywhere
Currently, min capture analysis results are written to MaybeTypeckResults, which is a transient data structure within rustc_typeck.

For the results to be available post-typeck they need to be written out to TypeckResults.

The writeback stage in `rustc_typeck/src/check/writeback.rs` needs to be updated to add the analysis results to the final results table.

### PR DESCRIPTION
**Pull request by jenniferwills**
Removed all uses of closures_captures and upvar_capture and refactored code to work with closure_min_captures. This also involved removing functions that were no longer needed like the bridge.

Closes https://github.com/rust-lang/project-rfc-2229/issues/18
r? @nikomatsakis

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]
Hmm, this had some subtle bits in it! But I think it's correct.

**Comment by nikomatsakis** [comment]
r=me, but I'd like to see the hashing change

@bors delegate+

File: compiler/rustc_middle/src/ty/context.rs [hunk 1]
```diff
@@ -693,7 +676,7 @@ impl<'tcx> TypeckResults<'tcx> {
 impl<'a, 'tcx> HashStable<StableHashingContext<'a>> for TypeckResults<'tcx> {
     fn hash_stable(&self, hcx: &mut StableHashingContext<'a>, hasher: &mut StableHasher) {
         let ty::TypeckResults {
-            hir_owner,
+            hir_owner: _,
```
[the remark this answers is not in the document]
Comment (arora-aman) [reply compiler/rustc_middle/src/ty/context.rs:679]: I'd assume it still does get hashed, upvar_capture_map got replaced by closure_min_capture map which contains places which represent their start using an UpvarId. These upvar ids are the same as those in the upvar_capture_map, they are now associated with other information like projections, types etc.

**Comment by rylev** [comment]
@jenniferwills @nikomatsakis After merging [another performance run](https://perf.rust-lang.org/compare.html?start=9f4bc3ead43a57783d8abea2fa6931a6736f3490&end=cebc8fef5f4391a9ed8e4c1dc566a6c5824e2901&stat=instructions:u) was done on this PR which shows regressions larger than 1%, meaning we might want to look into it. The regression is coming in incremental builds in the `encode_query_results_for` query which explains why this is impacting incremental builds. I personally don't see what changes could have caused a regression like this though. Any thoughts?

**Comment by nikomatsakis** [comment]
@rylev hmm, offhand no -- is this doing more encoding than previously? I'd actually expect *less* encoding work, since we removed some fields I think.

**Comment by arora-aman** [comment]
Only thing that has changed since the first perf run is that now we explicitly hash the `hir_owner` defid, this was previously being hashed part of the entries in the `upvar_capture_map`. I'd assume that added the extra work since the last run, though I'm not sure if that should be that significant. As @nikomatsakis said, overall we are still doing less encoding so I'm not sure why there is negative impact :/

**Pull request merged**